### ISSUE TITLE
Allow OAuth access to the "devices" API

### ISSUE THREAD

**Issue by rfk** [mozilla/fxa-auth-server#2547]
See also accompanying mailing-list discussion at https://groups.google.com/a/mozilla.com/forum/#!topic/app-services-internal/ePyK3SFT4Dg

We're trying to more to a world where different apps can have different levels of access to your account, managed via OAuth tokens and scopes.  One big thing that OAuth clients can't do right now is access the "devices" API to:

* Customize their display-name
* Register for push notifications about account status change
* Use the new "device commands" feature for things like send-tab

As noted in the above mailing-list thread, the devices API is pretty Firefox-centric - it assumes a world of similar-looking peers that all have full access to the user's Firefox Account.  I expect we'll probably need a few iterations of refactoring to move from where we are today, to a model that "fits" in an OAuth-style world.  But let's see if we can make the first such step.

Can we allow holders of OAuth tokens with the "https://identity.mozilla.com/apps/oldsync" scope to access the devices API in a similar manner to their sessionToken-holding peers?

The access-control is the easy part - we already know how to accept either a sessionToken or an OAuth token with specific scopes, from e.g. the `/account/profile` API.  But we'll have to figure out the following questions:

* How can we have a device record without a corresponding sessionToken?  There's some initial discussion of this in https://github.com/mozilla/fxa-auth-db-mysql/issues/378, but it'll also require some changes in this repo to deal with the redis sessionToken cache, last-access time tracking, and user-agent info tracking.

* How do we figure out what is the "current device" for a given request, a concept needed in `/account/device/commands` and probably some other places as well.  Currently we use the 1:1 mapping from sessionTokens to devices records to determine this.  The current suggestion is that we create a mapping from OAuth refresh tokens to device records and look it up via that, but maybe there are other options?

* What happens when you `/account/device/destroy` a device record for an OAuth client?  I think the expected behaviour would be that all of its OAuth tokens get destroyed, but that means we've got to reach over into fxa-oauth-server as part of destroying a device record.

* Do we need some notion of "placeholder" OAuth devices.  I'm hopeful that this won't be necessary, because OAuth tokens already show up properly in the "devices and apps" view.

* Will existing sync clients get confused?  If an OAuth app shows up in the list of devices on your account, will existing Firefox installs think it's another Firefox and offer to e.g. send tabs to it, despite the fact that it probably can't receive them yet?

@philbooth given your experience with the current devices API, I'd be really interested in your take on all this.

**Comment by rfk** [comment]
My strawman of a minimal set of necessary changes, including some issues already linked above:

* Change oauth-server to track what refresh token was used to create a given access token, and to report that in the response when verifying a token.
* Create some sort of backchannel whereby the auth-server can instruct the oauth-server to revoke a refresh token (or a set of tokens?).
* Remove the 1:1 mapping between device records and sessions; instead create two auxilliary mappings so that a device record can have *either* an associated sessionToken, or an associated refresh token (or set of tokens?).  This will probably be pretty hairy...
* Add oauth as an optional auth strategy for the device routes, requiring the "oldsync" scope.
* Change the logic for "what is the current device?" to work with OAuth tokens, by learing the corresponding refresh token from oauth verification, then looking up the device record for that refresh token.
* Change device destruction to also revoke the refresh token(s) associated with the device record.

That's...well, it's a lot of complexity, and it really smudges away the distinction between auth-server and oauth-server by introducing a pretty fundamental coupling between their data models.  But I think we're going to need to navigate this issue *somehow*, sooner or later, so maybe this is as good a place as any to start.

Pushback is definitely welcome though! :-)

**Comment by philbooth** [comment]
> @philbooth given your experience with the current devices API, I'd be really interested in your take on all this.

At a high level I think it all makes sense fwiw. (I did read @eoger's email about it too, should have replied there sorry)

Thinking specifically about the db change, I was wondering whether you could get to the same end result with less upheaval by just defining a new `oauthDevices` table and then reading from both of them when fetching a user's device list. That means an extra query for fetching of course but at least they can run concurrently. But maybe I haven't thought this through properly because I'm not clear how it would work with the "current device" logic.

Anyway, whatever, there's nothing you or @eoger have written that I'm against in principle!

> ...it really smudges away the distinction between auth-server and oauth-server...

Yeah, I know you're a long-time advocate of bringing those two servers together. That definitely makes more sense to me reading this thread!

**Comment by vladikoff** [comment]
from mtg: new approach in the works by getting a sessionToken out of the OAuth flow instead

**Comment by shane-tomlinson** [comment]
See https://docs.google.com/document/d/1cbz3pZHGTjJwLoDqZyjNTyzyE1IvEeDE_-uEWrH-g9g/

**Comment by vladikoff** [comment]
from mtg: planning and prototyping at this stage

**Comment by vladikoff** [comment]
from mtg: renamed to `instances`

**Comment by vladikoff** [comment]
from mtg: @eoger  and @rfk should this remain in 126 or be moved to 127?

**Comment by rfk** [comment]
I'd like to make as much progress as we can on this in 126.

**Comment by shane-tomlinson** [comment]
I’m really struggling to come to grips with the need for this API when we already have both the devices and sessions APIs. Is this an attempt at consolidating both? Neither [the feature doc](https://docs.google.com/document/d/1pbyOOySw76R-OTPJiX4gW3Izlp-plA7T7Jlt-wOAKwQ/edit) nor the related PRs present a problem statement or analysis that explains why the current APIs are insufficient and can't be made to be sufficient. Sure, "devices" is a confusing name, and already having 2 APIs is confusing, but so is adding yet another API. What's even more confusing is the new API accepts 3 distinct types of tokens. Is functionality missing that can't be added to the existing APIs? Can the existing APIs not be made to work in an OAuth world? Is there something inherently broken in the existing APIs that can't be made to work for our new use cases? Is the problem we want to work in a purely OAuth world and the existing APIs are on the old "auth-server" side of things?

What worries me most is the security and maintenance burden of supporting all of these APIs for the foreseeable future. A full migration path hasn't been defined, i.e, at what point do we decide the old APIs are no longer needed and we can stop supporting them? Even if we convert all existing consumers of the current APIs to use this new API, we know from [experience that deprecating support for old browsers is a multi-year process](https://github.com/mozilla/fxa-content-server/issues/5651). The existing APIs aren't going anywhere for the next several years, no matter how ambitious our aspirations.

I'm also afraid of hacks like retrofitting client instances onto devices and devices onto client instances. That seems like madness, something that will cause us to blow off both feet.

I'm adding a list of questions that @eoger, @vladikoff and I went through, many of which I am still unclear about:

* Why not just use the device registration API by extending it to accept OAuth refresh tokens?
* How are refresh tokens w/o an associated instance ID handled? Are existing refresh tokens w/o an instanceID ignored? Is an instanceID backfilled somehow? Is there a correction mechanism? If so, what does that look like?
* What is the Firefox migration story?
  * When can we get rid of the old stuff?
* Does the content server need to change, even if there is a mapping layer that converts devices to client instances and client instances to devices?
* What happens if one sessionToken has multiple refresh tokens associated with it? Is there de-duplication?
* What happens if an OAuth client never requested a refresh token?
* Say an existing OAuth relier who has users with existing access tokens (but no refresh tokens) decides they want to register a push endpoint, how do they do so? Does the user have to sign in again?
* How is the implicit grant flow where an access token is returned instead of a code? Will a refresh token be returned too? If not, how will these instances be managed?
* Why a refresh token instead of an access token? Is it because access tokens expire? Since refresh tokens are not guaranteed to exist, it seems like any endpoint that accepts a refresh token should also accept an access token.
* Requiring reliers to remember and pass an instanceID seems like too much of a burden. We already ask them to keep track of refresh and access tokens. Why can't we be trusted to maintain that mapping?
* What risks do we introduce by adding this new API? 
* What risks are there in keeping the existing API in parallel?
* What risks are there if we expand upon the existing API only?
* If we go ahead with this change, what is most likely to cause it to be derailed?

If we decide a new unified API really is the One True Way (TM), what is the smallest possible set of additional functionality that is needed? 3800 line diffs touching 58 files is.... big. Can that thing be broken up?

**Comment by rfk** [comment]
> If there is a 1:1 correspondence between an instanceID and a refresh_token,
> and their lifetimes are the same, why both?

The `refresh_token` is supposed to be a secret;  there needs to be some way of identifying the corresponding client without knowing what its `refresh_token` is and being able to impersonate that client.

**Comment by eoger** [comment]
This is a 6-hands answer to the very good feedback @shane-tomlinson has given:

> Why not just use the device registration API by extending it to accept OAuth refresh tokens?

Several reasons:
- the current devices API has some deficiencies (create versus update) that makes the registration process [more difficult than it should be](https://searchfox.org/mozilla-central/rev/8f89901f2d69d9783f946a7458a6d7ee70635a94/mobile/android/services/src/main/java/org/mozilla/gecko/fxa/devices/FxAccountDeviceRegistrator.java#334-396). The new endpoints simplify the API surface for “registration” and reuses prior art from FxA Commands. The close relationship with OAuth and refresh_tokens help us enforce integrity and order.
- the existing API is confusing and does not fit well in the “Applications” context. It was okay for Firefox to integrate with the devices API, but we cannot build new products without struggling with this API. “Devices” used to be a good term in the past where Mozilla was shipping only one application (Firefox) using the device registration API per device, therefore 1 app ~= 1 device. This is not true anymore with our multi-app strategy on mobile and is only adding more confusion as we are not designating a physical device. 
- the current device registration API is heavily [tied to session tokens and sessions](https://github.com/mozilla/fxa-auth-server/blob/5714695807892b086c74906515c85277dc03e61f/lib/db.js#L505-L535).
- overloading / extending the existing API could result in worse technical debt in the server code and database.

Sure we could probably shoehorn Instances into the devices API, however we believe the upsides and the clean separation between legacy and new is worth the trouble. It will also make it easier in the future to add specific "instances only" features/endpoints. 

We could think of this as Devices 2.0, except we renamed the API name as well. We learned from mistakes and decided to push forward on a more usable API. 

> How are refresh tokens w/o an associated instance ID handled? Are existing refresh tokens w/o an instanceID ignored? Is an instanceID backfilled somehow? Is there a correction mechanism? If so, what does that look like?

Newly refresh tokens would always have an instance ID associated with them.
“Pre-PR” refresh tokens without an associated instance ID will still show up as “empty” instances, just like any newly created refresh token with no associated metadata.
We should also “backfill”: there are several options here, either using a SQL script or have the server itself generate the Ids during runtime. For example, when an InstanceId is missing for some refresh_token (old token) then servers will generate one for that token when it is requested.

Finally, whenever we need to list the instances in an app, then we would use the properties, such as  `availableCommands` to filter out Instances that are not useful. For example, Send Tab in Fenix will only display a list of Instances where `send-tab` is available.

> If there is a 1:1 correspondence between an instanceID and a refresh_token, and their lifetimes are the same, why both? The only reason I can think of is their lifetime is not actually the same.

If a refresh token was generated before the Instances API then it won’t have an Instance Id.
The separation is here to make these concepts future proof and separate the concerns. It allows us to make sure the Instances do not have a hard dependency on the refresh_token id. We could make changes in the future where the type of tokens will change or we will need ways to reset/refresh. Instances with a different refresh_token.
Overloading the refresh_token with additional non-OAuth information is not beneficial.
[rfk] The refresh token is supposed to be a secret shared only between the client instance and the FxA servers, so if we want other clients to be able to refer to that instance without being able to impersonate it, then we need a stable non-secret identifier.

> Requiring reliers to remember and pass an instanceID seems like too much of a burden. We already ask them to keep track of refresh and access tokens. Why can't we be trusted to maintain that mapping?

When we designed instances, we imagined an instance ID being shared between multiple refresh tokens and clients explicitly knowing about them (e.g. for corrections on old tokens etc).
With a 1:1 refresh_token:instance_id mapping, there is no need do that anymore and we should update the PRs.

> What is the Firefox migration story? When can we get rid of the old stuff?

Firefox iOS is on its way of integrating our components next year, and also has very short tail of older versions seen in the wild.
Fennec should take the problem of our hands when it’s EOLed.
Desktop is a bit more problematic because of ESRs and the fact that we have no concrete plans at the moment of integrating the fxa component in there. However that’s something I’d like to work on in 2019.

> Does the content server need to change, even if there is a mapping layer that converts devices to client instances and client instances to devices?

Probably small changes on the “devices and apps” view to add deduplication.

[rfk] Rather than having content-server do more work, I think this is an opportunity to have it do less.

Right now it calls the /sessions endpoint on auth-server to get all the sessionToken clients, and then calls /client-tokens on the oauth-server to get all the OAuth clients.

I think this is an opportunity to add a single consistent "get all the clients connected to the account" endpoint ad simplify that logic on the content-server.

> What happens if one sessionToken has multiple refresh tokens associated with it? Is there de-duplication?

Each refresh token would be a new “instance”, these changes do rely on reliers being well-behaved. Note that a badly-behaved app could also do this today with sessionTokens, accidentally creating multiple device records for a single installed app.

> What happens if an OAuth client never requested a refresh token?

A relier must switch to start using refresh tokens to get this functionality. If they don’t then it’s business as usual for them.

[rfk] IIUC these would typically be on-public clients like AMO or Pocket, for whom I don't think there's any reason to want to use these APIs.  They might like to register a push callback in future, but would probably want to use a single callback associated with their client_id rather than a separate callback for every user that has signed in.

> Say an existing OAuth relier who has users with existing access tokens (but no refresh tokens) decides they want to register a push endpoint, how do they do so? Does the user have to sign in again?

Yes, the user would have to sign in again; that relier's access token has probably expired anyway.

[rfk] I don't believe we have any such reliers in practice, do we? (See above comments about AMO and Pocket).
 
> How is the implicit grant flow where an access token is returned instead of a code? Will a refresh token be returned too? If not, how will these instances be managed?

[rfk] AFAIK we don't currently support what the RFC calls the "implicit flow". We have something similar where sessionToken-holding clients can issue access_tokens to themselves, but will already show up as sessionToken-holding devices rather than OAuth instances.
Returning a refresh token in the implicit flow is explicitly forbidden by the OAuth spec RFC.

> Why a refresh token instead of an access token? Is it because access tokens expire? Since refresh tokens are not guaranteed to exist, it seems like any endpoint that accepts a refresh token should also accept an access token.

As for now if you want the feature you got to get a refresh token, because they are long lived.

[rfk] If it were an access token, what scope would we check for?  Would other services that the client has been using (and which have thus received OAuth access_tokens from that client) be able to access its instance metadata?

[rfk] Broadly, the refresh_token is for authenticating the client instance itself to the FxA servers, while access tokens are for accessing services on behalf of the user.

> What risks do we introduce by adding this new API?

[rfk] Unexpected server load issues if we forget something that we previously resolved for the devices API, such as the redis-level caching for frequently-changing info.
[rfk] Unexpected security/privacy issues, such as accidentally leaking info about what apps a user is using, to reliers that should not be able to access such info.

> What risks are there in keeping the existing API in parallel?

Have to maintain compat with older clients and new apps. Need to migrate the clients as well.
The “translation” layers (instances as devices, devices as instances) could bite us in the future when we do bigger changes.

> What risks are there if we expand upon the existing API only?

[vladikoff] see responses in q1 and q3.

[rfk] I think we could do this in the interests of keeping changes small.  We risk ongoing confusion about what a "device" means from our clients, e.g. expecting that multiple apps on the same phone would share a single device id and record.  We could paper over that in our client-side APIs by calling it something else.

> If we go ahead with this change, what is most likely to cause it to be derailed?

Database issues, such as migration of older tokens. 

Server load spike when a bunch of new clients come online and start using it 6 months from now when we're not paying attention to it anymore (e.g. Fenix release, or migrating Desktop Firefox to use it).  This happened to us previously with the devices API rollout.

**Comment by rfk** [comment]
> 3800 line diffs touching 58 files is.... big. Can that thing be broken up?

I also think this is the key to finding our way forward here, looking for incremental pieces of this that we can agree on and land independently.

**Comment by shane-tomlinson** [comment]
> Try to update this repo to use the new API from (2) and add oauth-authenticated access to the existing devices api. This may or may not actually work out in practice, and we don't want it to become a rabbit-hole.

This.... I realize this isn't the big "JUMP" that was proposed, but makes me much more comfortable. If it doesn't work out, then jump!

**Comment by rfk** [comment]
IIUC, the stages in https://github.com/mozilla/fxa-auth-server/issues/2547#issuecomment-443104884 are more-or-less what we settled on during discussion in Mozlando, I'll prioritize (2) this week.

**Comment by rfk** [comment]
@shane-tomlinson so IIUC this is the "root bug" for the oauth/send-tab work. I'm going to take the liberty of assigning it to you for further comment and triage on the path forward here, based on conversations over the past couple of days.

**Comment by shane-tomlinson** [comment]
I've been staring at the auth, oauth and auth-mysql, and several PRs across them for over a day and feel like I have a better grasp on this. I'm even more convinced that re-using the existing devices API will give us the best chance of success for the fast approaching Fenix release. Client instances is a great idea, but it's big. Bigger than I am comfortable committing our team to for Fenix.

I think the least amount of work will be to allow the existing devices API to be authenticated using an OAuth refresh token. 

🤢

Maybe. But so is two concurrent APIs that largely do the same thing.

There is some trickiness to this approach because 1) We'll have to ensure the refreshToken is valid and we've never checked refreshTokens from the auth-server before, only access tokens, 2) the devices API contains a lot of logic that relies upon data fetched from the sessionToken, e.g., all of the device info, and 3) Whenever the user deletes a device backed by a refreshToken, they'll expect the refreshToken to be invalidated. There is probably more that I'm not aware of since I'm still building a mental model of all the pieces, any additional info is welcome.

Performing operations with refreshTokens means we need to get access to functionality that is traditionally done on the oauth-server. I see a couple of ways to do this:

1. Expand the [oauthdb.js abstraction](https://github.com/mozilla/fxa-auth-server/blob/1013228206ac9326e8a2d5c976998ac730a06884/lib/oauthdb.js) to call routes on the oauth server. 
2. Get a reference to the OAuth server's db from within the auth server. We haven't done this in the past, and it might feel like we are breaking all kinds of boundaries, but isn't simplicity like this the reason we want to merge the auth & oauth servers?

If we go with expanding the oauthdb.js abstraction, we'll need at least one new route on the oauth server that is suitable for verifying refreshTokens. The /token endpoint *could* be used for this, but an accessToken is always created when presenting the refreshToken, we don't want that. Luckily the [OAuth token introspection spec](https://tools.ietf.org/html/rfc7662#section-2.2) was made just for this. Instead, I'd add a new route, `/introspect` that accepts a refreshToken and returns a bunch of data, including whether the token is valid and the uid associated with the refreshToken. This route would need some guards to mitigate iteration attacks, perhaps require either a client_id/client_secret or because the auth-server would be making these requests, allowed a shared secret between the auth/oauth server like in [@eoger's instance metadata PR](https://github.com/mozilla/fxa-auth-server/pull/2715/files#diff-4b2b29526aa31e22736772de3312e190R11). We'll likely need to add the same auth-strategy to [POST /destroy](https://github.com/mozilla/fxa-auth-server/blob/1013228206ac9326e8a2d5c976998ac730a06884/fxa-oauth-server/lib/routes/destroy.js) and any other routes needed.

Either way, once we verify a refreshToken we can create an object that has the same shape as a sessionToken and is passed in place of a sessionToken. 

We have to be able to provide the same functionality to refreshToken backed devices as to sessionToken backed devices. Device records currently have a sessionToken ID associated with them allowing for operations like:

1. look up a sessionToken from the device
2. look up a device for a given sessionToken
2. delete a sessionToken when a device is deleted
3. delete a device when a sessionToken is deleted

These operations are possible because a device has a [sessionToken ID stored with it](https://github.com/mozilla/fxa-auth-db-mysql/blob/310d4b4f539b35a8ff5eb8a976328129eaefc72a/lib/db/mysql.js#L333). We pass in a sessionToken to [devices->upsert here](https://github.com/mozilla/fxa-auth-server/blob/1013228206ac9326e8a2d5c976998ac730a06884/lib/devices.js#L92).

We *could* pass in and store the refreshToken ID as the sessionToken ID, but that feels off. Can we guarantee that a refreshToken and sessionToken will never have the same ID? While it's not likely to happen, I don't want to bet the farm on that. The other thing that feels off is wherever we go from a deviceID back to a sessionToken/refreshToken for things like destroying a device, we'd have to look in both the sessionToken and refreshToken tables. That just seems painful.

Maybe instead we should either store the token type that backs the device and use the sessionTokenId column for both session and refresh token IDs, or we add a new refreshTokenId column.

Perhaps there is a chance here to enable functionality for the pairing flow, specifically sending push notifications by [creating placeholder devices whenever a refreshToken is created with the "oldsync" scope](https://github.com/mozilla/fxa-auth-server/issues/2880#issuecomment-458795866)

In any case, it feels like we *should* be able to make this work without being too awful. I started on a very [non-working proof of concept](https://github.com/mozilla/fxa-auth-server/compare/devices-with-refresh-token?expand=1) to try to figure out some of my thoughts. Maybe that could be a useful starting point, or at least any case provide a tiny bit of insight into what's going on in my brain.

**Comment by rfk** [comment]
Renaming the issue to reflect its reduced scope...

**Comment by vladikoff** [comment]
from a meeting with @rfk and @vladikoff

Here's some PRs we need to PR:

- [ ] feat(devices): devices API with refreshTokens - https://github.com/mozilla/fxa-auth-server/pull/2937
 - We will end up with refreshTokenId entries in devices table where sessionTokenId is NULL, which is fine. But now SendTabs works!
- [ ] Fix the Fenix Devices & Apps List
  - [ ] Current Device in the list
  - [ ] Signing Out stuff
  - [ ] Users to migrated
- [ ] Upgrade /token code exchange, placeholder device record, and notify on device on connected.  https://github.com/mozilla/fxa-auth-server/pull/2931
- [ ]  /authorization with sessionToken for Fenix migration https://github.com/mozilla/fxa-auth-server/pull/2932
- [ ] Replace the `/clients-tokens` on the content server with a single call to the auth server (to make the devices and apps list better).
- [ ] Fenix Migrations: 
  - [ ] Pop the refreshTokenId on the new column, into existing entry.
  - [ ] The app needs to reset its device record, Fenix should re-register push
- [ ] Upgrade the /token code exchange to create `sessionToken`s AKA "superscope code exchange"

**Comment by vladikoff** [comment]
@rfk, from chatting with shane:

* With some updates to migration plans users will have 2 browsers installed, if we want to keep old fennec (side by side installation), we should reconsider keeping the device name, instead creating a new one from oauth client name  (talked about session/duplicate in this case) - *side-by-side migration*

* and then we have a "replace migration", we might need to somehow indicate that we want the name kept the same

**Comment by vladikoff** [comment]
* With some updates to migration plans users will have 2 browsers installed,
    * session duplicate OR super scope api?

**Comment by vladikoff** [comment]
* side-by-side migration: compares package name, if `fenix`, then create a device record
* replace migration migration: compares package name, if `fennec` then pop the new refreshTokenId onto existing name.

**Comment by vladikoff** [comment]
* do want to do side-by-side migration? gives us a way to test our migration, we don't lose a signed in user, we can do it for the UX. but not many people will get to it.

**Comment by rfk** [comment]
@vladikoff as you probably saw from your bugmail, I tried to make sure we had proper issues on file for the items in your checklist in https://github.com/mozilla/fxa-auth-server/issues/2547#issuecomment-468482625; the only one I didn't get to was "superscope code exchange" because I feel like we should probably have an IRL discussion with @shane-tomlinson  about where we want to go with that one...

**Comment by shane-tomlinson** [comment]
> @vladikoff vladikoff closed this in 3b33f41 13 hours ago

\o/ well done @vladikoff, @eoger, and @rfk. This is epic.
